src/bin/ls.rs

```diff
@@ -51,46 +54,100 @@ fn list_dir(path: &str, long_format: bool, string: &mut String, stderr: &mut Std
                 entry_path.push_str(&entry);
 
                 let metadata = fs::metadata(entry_path).try(stderr);
-                string.push_str(&format!("{:>7o} {:>5} {:>5} {:>8} ", metadata.mode(), metadata.uid(), metadata.gid(), metadata.size()/1024));
+                if human_readable {
+                    string.push_str(&format!("{:>7o} {:>5} {:>5} {} ",
+                                             metadata.mode(),
+                                             metadata.uid(),
+                                             metadata.gid(),
+                                             to_human_readable_string(metadata.size())));
+                } else {
+                    string.push_str(&format!("{:>7o} {:>5} {:>5} {:>8} ",
+                                             metadata.mode(),
+                                             metadata.uid(),
+                                             metadata.gid(),
+                                             metadata.size()));
+                }
             }
             string.push_str(entry);
             string.push('\n');
         }
     } else {
         if long_format {
-            string.push_str(&format!("{:>7o} {:>5} {:>5} {:>8} ", metadata.mode(), metadata.uid(), metadata.gid(), metadata.size()/1024));
+            if human_readable {
+                string.push_str(&format!("{:>7o} {:>5} {:>5} {} ",
+                                         metadata.mode(),
+                                         metadata.uid(),
+                                         metadata.gid(),
+                                         to_human_readable_string(metadata.size())));
+            } else {
+                string.push_str(&format!("{:>7o} {:>5} {:>5} {:>8} ",
+                                         metadata.mode(),
+                                         metadata.uid(),
+                                         metadata.gid(),
+                                         metadata.size()));
+            }
         }
         string.push_str(path);
         string.push('\n');
     }
 }
 
+fn to_human_readable_string(size: u64) -> String {
+    if size < 1024 {
+        return format!("{}", size);
+    }
+
+    let sizef = size as f64;
+    let units = ["", "K", "M", "G", "T", "P", "E"];
+
+    let digit_groups = (sizef.log10() / 1024f64.log10()) as i32;
```

Comment:
I think that it would be much better to use `u64::leading_zeros` to find `log2` of the value. It should be much faster than counting logarithm.